felix/daemon/daemon.go

```diff
@@ -322,16 +322,18 @@ configRetry:
 		typhaDiscoverer = createTyphaDiscoverer(configParams, k8sClientSet)
 		// Wireguard can block connection to Typha, add a post-discovery hook to detect and resolve any
 		// interactions.  (This will be a no-op if wireguard has never been turned on.)
-		typhaDiscoverer.AddPostDiscoveryFilter(func(typhaAddresses []discovery.Typha) ([]discovery.Typha, error) {
-			// Perform wireguard bootstrap processing. This may remove wireguard configuration if wireguard
-			// is disabled or if the configuration is obviously broken. This also filters the typha addresses
-			// based on whether routing is obviously broken to the typha node (due to wireguard routing
-			// asymmetry). If all typha instances would be filtered out then we temporarily disable wireguard
-			// on this node to allow bootstrap to proceed.
-			log.Info("Got post-discovery callback from Typha discoverer; checking if we need to " +
-				"filter out any Typha addresses due to Wireguard bootstrap.")
-			return bootstrapWireguardAndFilterTyphaAddresses(configParams, v3Client, typhaAddresses)
-		})
+		if configParams.WireguardEnabled || configParams.WireguardEnabledV6 {
+			typhaDiscoverer.AddPostDiscoveryFilter(func(typhaAddresses []discovery.Typha) ([]discovery.Typha, error) {
+				// Perform wireguard bootstrap processing. This may remove wireguard configuration if wireguard
```

Comment:
> This may remove wireguard configuration if wireguard is disabled

I don't think we can gate this entire function on wireguard enablement, because it appears to clean up wireguard configuration if Wireguard is disabled - which means we still need to run some part of this even if Wireguard isn't enabled (because it may have been in the past)

Reply:
Namely, this bit of logic: https://github.com/projectcalico/calico/blob/a4f1a953290119f0827c3ac6021c68ef65220950/felix/wireguard/bootstrap.go#L207-L211

Which AFAICT, doesn't interact with Wireguard itself? Have you confirmed that this function actually loads the wireguard module when wireguard is Disabled in Calico?  

AFAICT, it just does the equivalent of this:

- `ip link del <wireguard interface>`
- `kubectl edit node` to remove the public key annotation.

and then returns

Neither of which should load the wireguard module?

Reply:
You're right. I've chased places where `wgctrl.New()` is invoked which is loading wireguard module, and overlooked part where bootstrapping of wireguard is just clean configs when WG is not enabled. My mistake, I've fixed the code